Streaming files as partial content

AKazm opened the pull request:
Better support for streaming files by returning `206 Partial Content` when a range header is supplied with the correct content. Required when streaming to [some browsers](https://developer.apple.com/library/archive/documentation/AppleApplications/Reference/SafariWebContent/CreatingVideoforSafarioniPhone/CreatingVideoforSafarioniPhone.html#//apple_ref/doc/uid/TP40006514-SW6).

AKazm commented:
I've noticed two bugs in there:

- `asByteBufferBounds()` returnend incorrect values (fixed)
- Incorrect parsing of range headers; fixed that, too.

grosch commented on Sources/Vapor/HTTP/Headers/HTTPHeaders+ContentRange.swift:
```diff
@@ -0,0 +1,233 @@
+    case bytes
+    case custom(value: String)
+    
+    public func asHeaderString() -> String {
```
Can this be renamed `serialize()` to be consistent with many areas of Vapor?

AKazm replied:
Done.

grosch commented on Sources/Vapor/HTTP/Headers/HTTPHeaders+ContentRange.swift:
```diff
@@ -0,0 +1,233 @@
+        let (unitStr, rangesStr) = (splitResult[0], splitResult[1])
+        let ranges = rangesStr.split(separator: ",", omittingEmptySubsequences: false)
+            .map { it in RequestedRange.from(requestStr: it) }
+        if ranges.filter({ $0 == nil }).count > 0 {
```
Please use `!ranges.filter({ $0 == nil}).isEmpty` as it's more efficient.

AKazm replied:
Done.

AKazm commented:
Although it would somewhat obfuscate the method's purpose, it's fine by me, since consistency is important, too.

grosch commented on Sources/Vapor/HTTP/Headers/HTTPHeaders+ContentRange.swift:
```diff
@@ -0,0 +1,233 @@
+        self.ranges = ranges.compactMap { $0 }
+    }
+    
+    func asHeaderString() -> String {
```
Please use `serialize()`

AKazm replied:
Done.

grosch commented on Sources/Vapor/HTTP/Headers/HTTPHeaders+ContentRange.swift:
```diff
@@ -0,0 +1,233 @@
+        self.unit = lowerCasedUnit == "bytes"
+            ? RangeUnit.bytes
+            : RangeUnit.custom(value: lowerCasedUnit)
+        guard let contentRange = ResponseContentRange.from(responseStr: rangeStr) else {
```
Please move this to the line right after you assign `rangeStr`

AKazm replied:
I'm absolutely don't know what you mean by that. Which line should be moved?

grosch replied:
You assign `rangeStr` on line 64.  So move the entire guard block to right after line 64.  No need to set `self.unit` if you're going to leave the method.

AKazm replied:
Done.

AKazm replied:
Done.

grosch commented on Sources/Vapor/HTTP/Headers/HTTPHeaders+ContentRange.swift:
```diff
@@ -0,0 +1,233 @@
+        self.range = contentRange
+    }
+    
+    func asHeaderString() -> String {
```
Rename to `serialize()`

AKazm replied:
Done.

grosch commented on Sources/Vapor/HTTP/Headers/HTTPHeaders+ContentRange.swift:
```diff
@@ -0,0 +1,233 @@
+    public static func from<T>(responseStr: T) -> ResponseContentRange? where T : StringProtocol {
+        let ranges = responseStr.split(separator: "-", omittingEmptySubsequences: false)
+            .map({ $0.trimmingCharacters(in: .whitespacesAndNewlines) })
+        if ranges.count > 2 {
```
Swap to a `guard` please.

AKazm replied:
Done.

grosch commented on Sources/Vapor/HTTP/Headers/HTTPHeaders+ContentRange.swift:
```diff
@@ -0,0 +1,233 @@
+        return .withinWithLimit(start: start, end: end, limit: limit)
+    }
+    
+    public func asHeaderString() -> String {
```
`serialize()` please

AKazm replied:
Done.

grosch commented on Sources/Vapor/HTTP/Headers/HTTPHeaders+ContentRange.swift:
```diff
@@ -0,0 +1,233 @@
+        return .within(start: start!, end: end!)
+    }
+    
+    public func asHeaderString() -> String {
```
`serialize()` please :)

AKazm replied:
Done.

grosch commented:
Please add unit tests as well.

AKazm commented:
- renamed `asHeaderString()` to `serialize()`
- added tests for serialization of ranges and content ranges
- added tests for deserialization of ranges and content ranges
- resolved all code style issues mentioned by @grosch (except for "Please move this to the line right after you assign `rangeStr`", since I seriously don't get it :()

grosch commented on Tests/VaporTests/HTTPHeaderTests.swift:
```diff
@@ -2,6 +2,49 @@
+        XCTAssertEqual(RangeHeader(requestHeader: "bytes=200-1000, 2000-6576, 19000-, -500"), range)
+    }
+    
+    func testContentRangeSerialization() throws {
```
This is more just a style thing, so feel free to ignore, but this is really three separate tests, and so should be three separate test methods.  Same with `testContentRangeDeserialization`

AKazm replied:
Never did something like this before, to be honest.

So I did read some of the tests before and since there are several cases of tests containing multiple `XCTAsserEqual`, I thought this would be fine, since each tests covers the same functionality in one isolated "area".

MrLotU commented on Sources/Vapor/HTTP/Headers/HTTPHeaders+ContentRange.swift:
```diff
@@ -0,0 +1,232 @@
+        }
+        let (unitStr, rangesStr) = (splitResult[0], splitResult[1])
+        let ranges = rangesStr.split(separator: ",", omittingEmptySubsequences: false)
+            .map { it in RequestedRange.from(requestStr: it) }
```
```suggestion
            .map(RequestedRange.from)
```

MrLotU commented on Sources/Vapor/HTTP/Headers/HTTPHeaders+ContentRange.swift:
```diff
@@ -0,0 +1,232 @@
+    
+    public static func from<T>(responseStr: T) -> ResponseContentRange? where T : StringProtocol {
+        let ranges = responseStr.split(separator: "-", omittingEmptySubsequences: false)
+            .map({ $0.trimmingCharacters(in: .whitespacesAndNewlines) })
```
Don't need the `()` around the closure here. 
```suggestion
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
```

MrLotU commented on Sources/Vapor/HTTP/Headers/HTTPHeaders+ContentRange.swift:
```diff
@@ -0,0 +1,232 @@
+            }
+            return .any(size: size)
+        }
+        guard let start = Int(ranges[0]) else {
```
I think we need a check on `ranges.count > 0` or `!ranges.isEmpty` otherwise this _could_ crash with an `Index out of bound`

AKazm replied:
`ranges.count` cannot be `0` because i'm not omitting empty subsequences (i think?!)

MrLotU replied:
Ah. My bad. You're correct
```
  1> "".split(separator: "/") 
$R0: [String.SubSequence] = 0 values
  2> "".split(separator: "/", omittingEmptySubsequences: false).map(String.init) 
$R1: [String] = 1 value {
  [0] = ""
}
```

MrLotU commented on Sources/Vapor/HTTP/Headers/HTTPHeaders+ContentRange.swift:
```diff
@@ -0,0 +1,232 @@
+        let ranges = requestStr.split(separator: "-", omittingEmptySubsequences: false)
+            .map({ $0.trimmingCharacters(in: .whitespacesAndNewlines) })
+        guard ranges.count <= 2 else { return nil }
+        let start = Int(ranges[0])
```
Same `ranges.count > 0` or `!ranges.isEmpty` required here 😃

AKazm replied:
same: `ranges.count` cannot be `0` because i'm not omitting empty subsequences (i think?!)

MrLotU commented on Sources/Vapor/HTTP/Headers/HTTPHeaders+ContentRange.swift:
```diff
@@ -0,0 +1,232 @@
+    public func asByteBufferBounds(withMaxSize size: Int) -> (offset: Int64, byteCount: Int) {
+        switch self {
+            case .start(let value):
+                let int64Value = NSNumber(value: value).int64Value
```
Specific reason for `NSNumber` here? Why not `Int64(value)`?

AKazm replied:
well, i've used mutliple programming languages in my lifetime, none of them offered a safe way to convert `Int` to `Int64` or the other way around using a constructor (throwing, crashing, ..)

so, this is just based on an educated guess (which often turns out to be wrong when it comes to Swift, since Swift is very exotic compared to most languages). If I'm wrong here, I'm happy to learn something new!

Joannis replied:
Int and Int64 are practically the exact same type (although the compiler won't recognize it as such) unless you're running on a 32-bit platform. But who does that, and why would you want to support 32 bit platforms?

So because Int equals Int64 you're safe to do so.

AKazm replied:
I know about the quality of those types, what was especially confusing to me was SwiftNIOs explicit offset for reading files, which is of type `Int64`. If Vapor does not support 32bit systems, then I'll change it to initializer invocations alter today!

Joannis replied:
I can imagine! NIO is a pretty broad and a _very_ correct implementation.

AKazm commented on Sources/Vapor/Utilities/FileIO.swift:
```diff
@@ -133,7 +136,15 @@ public struct FileIO {
+            response.headers.add(name: .accept, value: contentRange.unit.serialize())
+            let range = contentRange.ranges[0].asResponseContentRange(limit: fileSize)
+            response.headers.contentRange = ContentRangeHeader(unit: contentRange.unit, range: range)
+            (offset, byteCount) = contentRange.ranges[0].asByteBufferBounds(withMaxSize: fileSize)
```
I notice undefined behaviour here if the requested range exceeds the file size. Any proposals here?

0xTim replied:
I think a bad request would be appropriate in this case

Joannis commented on Sources/Vapor/Utilities/FileIO.swift:
```diff
@@ -133,7 +136,15 @@ public struct FileIO {
+        if let contentRange = contentRange {
+            response.status = .partialContent
+            response.headers.add(name: .accept, value: contentRange.unit.serialize())
+            let range = contentRange.ranges[0].asResponseContentRange(limit: fileSize)
```
Is it safe to assume `!contentRange.ranges.isEmpty`? If it would be `[]` then this code will cause a crash for sure.

AKazm replied:
It is, since `contentRange` is `nil` if `ranges.count` is `!= 1` (see lines 122-124)

Joannis replied:
Thanks, I definitely overlooked that

Joannis commented:
Thanks so much for this PR. I'm afraid I don't have a quick solution for your question. But a `.badRequest` error would be an option in my book.

AKazm commented:
One option would be to return 416, another would be to just return the whole file instead. Both options would be valid, I'd prefer a configurable solution with a reasonable default.

tanner0101 commented on Sources/Vapor/HTTP/Headers/HTTPHeaders+ContentRange.swift:
```diff
@@ -0,0 +1,232 @@
+    }
+}
+
+public struct RangeHeader : Equatable {
```
These types should ideally be nested inside of `HTTPHeaders`. See https://github.com/vapor/vapor/blob/master/Sources/Vapor/HTTP/Headers/HTTPHeaders%2BContentDisposition.swift#L20 for a good example to copy.

tanner0101 commented on Sources/Vapor/HTTP/Headers/HTTPHeaders+ContentRange.swift:
```diff
@@ -0,0 +1,232 @@
+    
+    var contentRange: ContentRangeHeader? {
+        get {
+            return contains(name: .contentRange)
```
This should be using the `parseDirectives` methods. This takes care of a lot of weird edge cases in HTTP header values automatically. 

Again HTTPHeaders+ContentDisposition is a good example to follow here: https://github.com/vapor/vapor/blob/master/Sources/Vapor/HTTP/Headers/HTTPHeaders%2BContentDisposition.swift#L20

AKazm replied:
Done, although I can't say I'd like it.

tanner0101 commented on Sources/Vapor/HTTP/Headers/HTTPHeaders+ContentRange.swift:
```diff
@@ -0,0 +1,232 @@
+
+public struct ContentRangeHeader : Equatable {
+    public let unit: RangeUnit
+    public let range: ResponseContentRange
```
Instead of an enum, what about something like:
```swift
public var range: Range<Int>?
public var size: Int?
```
Where `nil` on either is serialized as `*`.

AKazm replied:
This actually has been my first approach on doing this, but I wanted to eliminate optionals here. Has the advantage of avoiding a lot of code cluttering.

If you insist on this change (same goes for Request Range), please confirm this!

tanner0101 commented on Sources/Vapor/HTTP/Headers/HTTPHeaders+ContentRange.swift:
```diff
@@ -0,0 +1,232 @@
+
+public struct RangeHeader : Equatable {
+    public let unit: RangeUnit
+    public let ranges: [RequestedRange]
```
Could we use `Swift.Range` here instead?

AKazm replied:
This actually has been my first approach on doing this, but I wanted to eliminate optionals here. Has the advantage of avoiding a lot of code cluttering.

If you insist on this change (same goes for Response Content Range), please confirm this!

tanner0101 commented on Sources/Vapor/HTTP/Headers/HTTPHeaders+ContentRange.swift:
```diff
@@ -0,0 +1,232 @@
+    }
+}
+
+public enum RequestedRange : Equatable {
```
This could be nested in `HTTPHeaders.Range`

AKazm replied:
This actually has been my first approach on doing this, but I wanted to eliminate optionals here. Has the advantage of avoiding a lot of code cluttering.

If you insist on this change (same goes for Response Content Range), please confirm this!

tanner0101 commented on Sources/Vapor/Utilities/FileIO.swift:
```diff
@@ -92,7 +92,8 @@ public struct FileIO {
         }
-        return self.read(path: path, fileSize: fileSize.intValue, chunkSize: chunkSize, onRead: onRead)
+        return self.read(path: path, fromOffset: 0, byteCount: fileSize.intValue, chunkSize: chunkSize,
+                         onRead: onRead)
```
No "xcode" indentation please. Look at other files to see how long method names are indented.

tanner0101 commented on Sources/Vapor/Utilities/FileIO.swift:
```diff
@@ -133,18 +137,25 @@ public struct FileIO {
-
+        var offset: Int64 = 0
+        var byteCount: Int = fileSize
+        if let contentRange = contentRange {
```
I think it would be more readable to have the above variables as `let` and set the default values in an `else`.

tanner0101 commented on Sources/Vapor/HTTP/Headers/HTTPHeaders+ContentRange.swift:
```diff
@@ -0,0 +1,232 @@
+
+extension RequestedRange {
+    
+    public func asByteBufferBounds(withMaxSize size: Int) -> (offset: Int64, byteCount: Int) {
```
We should be fine to just use `numericCast` here when we need to convert Int -> Int64. I'd much rather do that than use `NSNumber` without looking more deeply into the implementation. Who knows what dark magic Foundation is doing. 

Either way, this extension would make more sense next to the FileIO code as a private method.

tanner0101 commented on Tests/VaporTests/HTTPHeaderTests.swift:
```diff
@@ -2,6 +2,49 @@
+        let rangeOfUnknownLimit = "bytes 0-1000/*"
+        let rangeWithLimit = "bytes 0-1000/1001"
+        XCTAssertEqual(ContentRangeHeader(responseHeader: anyRange),
+                       ContentRangeHeader(unit: .bytes, range: .any(size: 1000)))
```
"xcode" formatting here, too

tanner0101 commented on Sources/Vapor/HTTP/Headers/HTTPHeaders+ContentRange.swift:
```diff
@@ -0,0 +1,259 @@
+import Foundation
+
+public extension HTTPHeaders {
```
`public` keyword should go in front of each type definition / method.

tanner0101 commented on Sources/Vapor/HTTP/Headers/HTTPHeaders+ContentRange.swift:
```diff
@@ -0,0 +1,259 @@
+import Foundation
+
+public extension HTTPHeaders {
+        
```
No new line here.

tanner0101 commented on Sources/Vapor/HTTP/Headers/HTTPHeaders+ContentRange.swift:
```diff
@@ -0,0 +1,259 @@
+
+public extension HTTPHeaders {
+        
+    enum RangeUnit : Equatable {
```
No space before colon.

tanner0101 commented on Sources/Vapor/HTTP/Headers/HTTPHeaders+ContentRange.swift:
```diff
@@ -0,0 +1,259 @@
+        
+        public func serialize() -> String {
+            switch self {
+                case .bytes:
```
Indentation.

tanner0101 commented on Sources/Vapor/HTTP/Headers/HTTPHeaders+ContentRange.swift:
```diff
@@ -0,0 +1,259 @@
+            self.ranges = ranges
+        }
+        
+        public init?(httpHeaders headers: HTTPHeaders) {
```
`HTTPHeaders` helpers should follow a similar pattern. `HTTPHeaders.ContentDisposition` is a good example. There should be internal `init?(directives: [Directive])` and `directives() -> [Directive]` methods that are used by a public mutable property like `public var contentDisposition: ContentDisposition?`.

tanner0101 commented on Sources/Vapor/HTTP/Headers/HTTPHeaders+ContentRange.swift:
```diff
@@ -0,0 +1,259 @@
+        }
+        
+        public init?(requestHeader header: String) {
+            let splitResult = header.split(separator: "=", omittingEmptySubsequences: false)
```
This type of parsing should be handled by `parseDirectives` as it's non-trivial to get right. For example, your code here isn't taking into consideration quoted strings.

tanner0101 commented on Sources/Vapor/Utilities/FileIO.swift:
```diff
@@ -109,7 +115,8 @@ public struct FileIO {
     /// - returns: A `200 OK` response containing the file stream and appropriate headers.
-    public func streamFile(at path: String, chunkSize: Int = NonBlockingFileIO.defaultChunkSize) -> Response {
+    public func streamFile(at path: String, chunkSize: Int = NonBlockingFileIO.defaultChunkSize,
+                           mediaType: HTTPMediaType? = nil) -> Response {
```
No editor assisted indentation, please. In other words, never indent more than one tab-width.

tanner0101 commented on Sources/Vapor/Utilities/FileIO.swift:
```diff
@@ -92,7 +92,13 @@ public struct FileIO {
+            byteCount:
+            fileSize.intValue,
+            chunkSize: chunkSize,
+            onRead: onRead)
```
`)` on newline.

tanner0101 commented on Sources/Vapor/Utilities/FileIO.swift:
```diff
@@ -187,3 +208,18 @@ public struct FileIO {
+            case .start(let value):
+                return (offset: numericCast(value), byteCount: size-value)
+            case .tail(let value):
+                return (offset: numericCast(size-value), byteCount: value)
```
Spaces around operators.

tanner0101 reviewed (dismissed):
The latest changes look good, thanks! Most of my comments here are style nits. Please take a look at the other source files in this project for comparison and follow their style. 

More importantly though, header value parsing should be handled by the `HTTPHeaders.Directive` helpers. From what I can tell this should handle the Range and ContentRange header syntax. Simple string parsing is error prone and will not handle edge cases.

AKazm commented:
Will take a look at it on Saturday.

AKazm commented:
Done, @tanner0101!

tanner0101 commented:
@AKazm sorry for the late response. The latest updates look great, thank you! Would you mind adding some comments to the code? Public method docblocks and inline comments around terse code would be especially helpful. The header parsing tests are great. We should have at least one test ensuring the actual partial file reading works though. Ideally this would go through a real route. This file has some good examples: https://github.com/vapor/vapor/blob/master/Tests/VaporTests/FileTests.swift. Let me know if you need help writing that.

AKazm commented:
Comments won't be a problem, writing test code for streaming files is a thing I've never done before and since I'm kind of occupied for the next couple of weeks, any help on this would be appreciated.

tanner0101 commented:
@AKazm let me know when this is ready for another review. Thanks!

AKazm commented:
@tanner0101 Will take care of the tests on the weekend!

AKazm commented:
I'm sorry, but I simply couldn't do it this weekend, @tanner0101.

I didn't forget about it and I'll take care of it as soon as I can, probably on Tuesday afternoon, CEST.

nerdsupremacist commented:
@AKazm What's the status on this? I kinda need it right now.

I'll do a fork of this for now, please reach out if you need any help

AKazm commented:
It's a motivation for me that other people want this as well.

I promise to solve this during the wekeend. Time is a serious issue for me right now and I'm sorry for not having been able to continue here within the past few months.

**Any help is appreciated**.

0xTim commented on Sources/Vapor/HTTP/Headers/HTTPHeaders+ContentRange.swift:
```diff
@@ -0,0 +1,263 @@
+                }
+                return .any(size: size)
+            }
+            guard let start = Int(ranges[0]) else {
```
Are there any scenarios where `range.count` can be 0? If so we should add a check here

0xTim commented on Sources/Vapor/HTTP/Headers/HTTPHeaders+ContentRange.swift:
```diff
@@ -0,0 +1,263 @@
+            guard let start = Int(ranges[0]) else {
+                return nil
+            }
+            let limits = ranges[1].split(separator: "/", omittingEmptySubsequences: false)
```
Again, we're not checking the array size before accessing here. If the array can be of size one or two, the guard at the top should be changed to reflect that

siemensikkema replied:
One way to achieve that and maybe gain some code clarity is to switch over `ranges.count`. This also avoids recounting.

```swift
switch ranges.count {
case 1: ...
case 2: ...
default: // either 0 or >2 
    return nil
```

0xTim commented on Sources/Vapor/Utilities/FileIO.swift:
```diff
@@ -133,18 +146,28 @@ public struct FileIO {
+        if let contentRange = contentRange {
+            response.status = .partialContent
+            response.headers.add(name: .accept, value: contentRange.unit.serialize())
+            let range = contentRange.ranges[0].asResponseContentRange(limit: fileSize)
```
Again, not checking array size here before accessing

0xTim reviewed (changes requested):
Thanks for doing this (and poking me!)

I think this is mostly good to go, just a few small things to clarify. Also, if you could update the PR comments as they'll become the release notes

siemensikkema commented on Sources/Vapor/HTTP/Headers/HTTPHeaders+ContentRange.swift:
```diff
@@ -0,0 +1,263 @@
+    /// Convenience for accessing the Content-Range response header.
+    ///
+    /// See https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Range
+    public var contentRange: ContentRange? {
+        get {
+            let directives = parseDirectives(name: .contentRange).flatMap { $0 }
+            return HTTPHeaders.ContentRange(directives: directives)
+        }
+        set {
+            if contains(name: .contentRange) {
+                remove(name: .contentRange)
+            }
+            if newValue == nil {
+                return
+            }
+            add(name: .contentRange, value: newValue!.serialize())
+        }
+    }
```
- Add a couple of `self`s to be in compliance with [the style guide](https://github.com/vapor/vapor/blob/master/.github/contributing.md)
- write getter in same style as for `range`
- use guard to avoid force unwrap
```suggestion
    public var contentRange: ContentRange? {
        get {
            return HTTPHeaders.ContentRange(directives: self.parseDirectives(name: .contentRange).flatMap { $0 })
        }
        set {
            if self.contains(name: .contentRange) {
                self.remove(name: .contentRange)
            }
            guard let newValue == newValue else {
                return
            }
            self.add(name: .contentRange, value: newValue.serialize())
        }
    }
```

siemensikkema commented on Sources/Vapor/HTTP/Headers/HTTPHeaders+ContentRange.swift:
```diff
@@ -0,0 +1,263 @@
+    ///
+    /// See https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Range
+    public var range: Range? {
+        get {
+            return HTTPHeaders.Range(directives: self.parseDirectives(name: .range).flatMap { $0 })
+        }
+        set {
+            if contains(name: .range) {
+                remove(name: .range)
+            }
+            if newValue == nil {
+                return
+            }
+            add(name: .range, value: newValue!.serialize())
+        }
```
- Add a couple of selfs to be in compliance with the style guide
- use guard to avoid force unwrap
```suggestion
        get {
            return HTTPHeaders.Range(directives: self.parseDirectives(name: .range).flatMap { $0 })
        }
        set {
            if self.contains(name: .range) {
                self.remove(name: .range)
            }
            guard let newValue == newValue else {
                return
            }
            self.add(name: .range, value: newValue.serialize())
        }
```

siemensikkema commented on Sources/Vapor/HTTP/Headers/HTTPHeaders+ContentRange.swift:
```diff
@@ -0,0 +1,263 @@
+        public static func from<T>(requestStr: T) -> HTTPHeaders.Range.Value? where T: StringProtocol {
+            let ranges = requestStr.split(separator: "-", omittingEmptySubsequences: false)
+                .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
+            guard ranges.count <= 2 else { return nil }
+            let start = Int(ranges[0])
+            let end = Int(ranges.count > 1 ? ranges[1] : "")
+            if start == nil && end == nil {
+                return nil
+            }
+            if start != nil && end == nil {
+                return .start(value: start!)
+            }
+            if start == nil && end != nil {
+                return .tail(value: end!)
+            }
+            return .within(start: start!, end: end!)
```
This is a way to avoid force unwraps:
```suggestion
            let count = ranges.count
            guard count <= 2 else { return nil }
    
            switch (count > 0 ? Int(ranges[0]) : nil, count > 1 ? Int(ranges[1]) : nil) {
            case (nil, nil):
                return nil
            case let (.some(start), nil):
                return .start(value: start)
            case let (nil, .some(tail)):
                return .tail(value: tail)
            case let (.some(start), .some(end)):
                return .within(start: start, end: end)
            }
```

siemensikkema commented on Sources/Vapor/HTTP/Headers/HTTPHeaders+ContentRange.swift:
```diff
@@ -0,0 +1,263 @@
+            }
+        }
+    }
+    
```
```suggestion
```

siemensikkema commented on Sources/Vapor/Utilities/FileIO.swift:
```diff
@@ -118,7 +129,9 @@ public struct FileIO {
             return Response(status: .internalServerError)
         }
-
+        let contentRange = request.headers.range?.unit == .bytes && (request.headers.range?.ranges.count ?? 0) == 1
+            ? request.headers.range!
+            : nil
```
```suggestion
        let contentRange = request.headers.range.map { range in
            guard range.unit == .bytes, range.ranges.count == 1 else {
                return nil
            }
            return range
        }
```

siemensikkema commented on Sources/Vapor/Utilities/FileIO.swift:
```diff
@@ -187,3 +212,18 @@ public struct FileIO {
+                return (offset: numericCast(start), byteCount: end + 1)
+        }
+    }
+    
```
```suggestion
```

siemensikkema reviewed (commented):
I left some suggestions to remove some force unwraps and to clarify the code here and there a bit.

0xTim commented:
I think this is finally ready - @siemensikkema @gwynne if you want to do a final review.

@AKazm is there anything else that needs addressing before we merge?

kylebrowning commented on Sources/Vapor/Utilities/FileIO.swift:
```diff
@@ -92,7 +92,14 @@ public struct FileIO {
+        return self.read(
+            path: path,
+            fromOffset: 0,
+            byteCount:
+            fileSize.intValue,
```
nit: extra newline?

AKazm replied:
> I think this is finally ready - @siemensikkema @gwynne if you want to do a final review.
> 
> @AKazm is there anything else that needs addressing before we merge?

Hello 0xTim,
please don't ask me for this. I've been in trouble since last year, I'm unfortunately occupied since my last activity here (since Corona gave our company many oppurtinities).

Nevertheless: I'm happy to see that there's some progress here made by other people since this is my first contribution to Open Source ever!

0xTim commented:
Confirmed I can play an MP4 in Safari so will merge, thank you everyone!

tanner0101 commented:
These changes are now available in [4.37.0](https://github.com/vapor/vapor/releases/tag/4.37.0)
